teps/draft-20200715-access-control-via-sa.md

```diff
@@ -0,0 +1,249 @@
+these different objects.
+
+This proposal will introduce means for providing finer grained control on which PipelineRun and TaskRun objects 
+can access various Pipeline, Task, and ClusterTask objects.
+
+**NOTE:** This [design doc](https://docs.google.com/document/d/1fOTZM7j9eH5qQHW0zvG82eRBrdnhdw2Z05d-vdlJph8/edit#heading=h.l0ysfwvdrqaw)
+serves as a superset of the scenarios addressed here, as well as a large amount of Kubernetes reference information and 
+links for concepts referenced here.  This document will attempt to minimize duplication of all that background.  But if 
+you are unfamiliar with a concept reference here, if that document can accelerate you research, please join tekton-dev
+if need be and bring it up.  
+
+## Motivation
+
+In the Kubernetes controller ecosystem, the notion of controllers creating objects on the behalf human users
+(via their creation of other objects) has resulted in a series of features inside Kubernetes around authorization
+(RBAC and admission controllers/plugins/webhooks among others) that allow administrators to restrict the amount 
+Kubernetes API interaction a controller can do on behalf of a given user.
+
+Similarly, in pre-existing CI/CD system (Jenkins for example), there are similar capabilities in place that allow
+administrators to read/write/execute access permissions on global Jobs/Pipelines (rough equivalent of cluster scoped 
+resources like ClusterTasks) and Jobs/Pipelines within Folders (rough equivalent of namespace scoped resources like 
+Tasks) for specific users or groups of users.    
+
+As part of both
+
+- Continuing the perception Tekton as a true, Kubernetes native CI/CD system, that leverages all the inherent 
+capabilities of Kubernetes
+- Reaching at least equal capability, compared to existing options, across the spectrum of concerns of a CI/CD system
+
+This proposal articulates a path to finer grained authorization/access control around the various Tekton concepts.
+
+### Goals
+
+One goal of this TEP is to define an approach for access control for references from one Tekton object to another in
+a manner that is natural for Kubernetes cluster administrators.
+
+A second goal of this TEP is introduce such access control in a manner that does not immediately disrupt existing 
+deployments of Tekton, given the free access to all approach that currently exists for these Tekton object to Tekton
+object references.
+
+### Non-Goals
+
+Other valid authorization related scenarios have been raised in various community discussions.  This proposal 
+does not attempt to address all of them.
+
+Among these various scenarios, there has also been discussion around introducing OPA (either via the admission webhook 
+for Kubernetes provided by the OPA org, or some other form of integration) into Tekton.  This proposal is not taking
+on how OPA would be integrated / leveraged by Tekton. 
```

Comment:
Can you elaborate on why using a tool like OPA isn't under consideration for evaluating this problem? I think this could be a really promising solution.

If what we're trying to do is restrict who has access to what Tasks / ClusterTasks, it feels like a small step from "service account B shouldn't be able to access ClusterTasks" to "service account B should only be able to access ClusterTasks A, B and C".

If we direct users to a solution like OPA for this use case (maybe even building a Tekton admission controller to help with this) we can offer a lot more flexibility.

Reply:
I would contend OPA was considered @bobcatfish since I listed it here as an alternative.

I'll respond by adjusting your question slightly.  Why was it not listed as the preferred solution?  My thoughts there were:

a) simplicity and solving this particular scenario
b) importance of solving this scenario sooner rather than later for users I see consuming tekton
c) uncertainly if we would ship OPA integration with only predefined rego policies for our scenarios of concern, vs. the potentially slippery slop of opening up any OPA integration to the user defining their own rego policies ... they are very flexible and powerful, but there will be an educational costs learning how to properly set up rego policies ... so an incremental step here

Certainly OPA has come up in several scenario discussions besides this one.  Have there been plans already that I am unaware at the board level around introducing OPA to tekton in general? 

Something of that nature feels like a separate project in and of itself as well.

And do we have folks in the tekton community already comfortable enough with OPA, gatekeeper, etc. to take on that effort?

Reply:
> Certainly OPA has come up in several scenario discussions besides this one. Have there been plans already that I am unaware at the board level around introducing OPA to tekton in general?

I believe that some folks at RedHat are looking into integrations between Tekton + OPA - @vdemeester might know more, not sure who else to point you toward ( @dlorenc maybe? )

> b) importance of solving this scenario sooner rather than later for users I see consuming tekton

I don't want to rush something in - I might need to understand better why we need something quickly here. My perspective is that via webhooks, users have the power to add whatever additional requirements they want, i.e. tekton doesn't need to add support for this in order for users to have it (which is one of the biggest wins of building on kubernetes in my opinion!), i.e. i dont see the hurry to get this functionality in.

I basically don't want to compromise the quality of our solution b/c we want something quickly:
a) if we think this is important enough that tekton needs to support it out of the box, id like to treat this with the importance it needs and approach it carefully, and potentially even encourage people to stop working on features to work on this instead
b) if this needs to be done quickly and we are planning to devote the time to it that it needs, whatever we do now  quickly should be designed to be removed

I don't see the need for (a) yet, but I might need to better understand the use cases that are motivating this.

Reply:
> > Certainly OPA has come up in several scenario discussions besides this one. Have there been plans already that I am unaware at the board level around introducing OPA to tekton in general?
> 
> I believe that some folks at RedHat are looking into integrations between Tekton + OPA - @vdemeester might know 

yep I'm familiar with that piece (and am involved with it as well behind the scenes) ... but I was curious what existed outside of Red Hat, given the amount of OPA references I see here and elsewhere under the broader tektoncd umbrella

thanks

more, not sure who else to point you toward ( @dlorenc maybe? )
> 

yeah let's see if @dlorenc chimes in

> > b) importance of solving this scenario sooner rather than later for users I see consuming tekton
> 
> I don't want to rush something in - I might need to understand better why we need something quickly here. My perspective is that via webhooks, users have the power to add whatever additional requirements they want, i.e. tekton doesn't need to add support for this in order for users to have it (which is one of the biggest wins of building on kubernetes in my opinion!), i.e. i dont see the hurry to get this functionality in.
> 
> I basically don't want to compromise the quality of our solution b/c we want something quickly:
> a) if we think this is important enough that tekton needs to support it out of the box, id like to treat this with the importance it needs and approach it carefully, and potentially even encourage people to stop working on features to work on this instead
> b) if this needs to be done quickly and we are planning to devote the time to it that it needs, whatever we do now quickly should be designed to be removed
> 
> I don't see the need for (a) yet, but I might need to better understand the use cases that are motivating this.